proposals/0177-program-runtime-abiv2.md

```diff
@@ -0,0 +1,156 @@
+scratchpad data. It must be updated when `sol_set_return_data` is called.
+
+### Per Instruction Serialization
+
+For each instruction the program runtime must prepare the following.
+This memory region starts at `0x600000000` and is readonly. It does not require
+any updates once serialized.
+
+- The instruction data: `&[u8]` which is composed of:
+  - Pointer to instruction data: `u64`
+  - Length of instruction data: `u64`
+- Programm account index in transaction: `u16`
+- Number of instruction accounts: `u16`
+- For each instruction account:
+  - Index to transaction account: `u16`
+  - Flags bitfield: `u16` (bit 0 is signer, bit 1 is writable)
+
+### Per Instruction Mappings
+
+A readonly memory region starting at `0x700000000` must be mapped
+in for the instruction data. It too does not require any updates.
+
+For each unique (meaning deduplicated) instruction account the payload must
+be mapped in at `0x800000000` plus `0x100000000` times the index of the
+**transaction** account (not the index of the instruction account). Only if the
+instruction account has the writable flag set and is owned by the current
+program it is mapped in as a writable region. The writability of a region must
+be updated as programs through out the transaction modify the account metadata.
+
+### Lazy deserialization on the dApp side (inside the SDK)
+
+With this design a program SDK can (but no longer needs to) eagerly deserialize
+all account metadata at the entrypoint. Because this layout is strictly aligned
+and uses proper arrays, it is possible to directly calculate the offset of a
+single accounts metadata with only one indirect lookup and no need to scan all
+preceeding metadata. This allows a program SDK to offer a lazy interface which
+only interacts with the account metadata fields which are needed, only of the
+accounts which are of interest and only when necessary.
+
+### Changes to syscalls
+
+The `AccountInfo` parameter of the CPI syscalls (`sol_invoke_signed_c` and
+`sol_invoke_signed_rust`) will be ignored if ABI v2 is in use. Instead the
+changes to account metadata will be communicated explicitly through separate
+syscalls `sol_set_account_owner`, `sol_set_account_lamports` and
+`sol_set_account_length`. Each of these must take a guest pointer to the
+structure of the transaction account (see per transaction serialization) to be
+updated and the new value as second parameter. In case of the pubkey parameter
+the guest pointer to a 32 byte slice is taken instead.
+
+### Changes to CU metering
+
+CPI will no longer charge CUs for the length of account payloads. Instead TBD
+CUs will be charged for every instruction account.
+
+## Impact
+
+This change is expected to drastically reduce the CU costs as the cost will no
+longer depend on the length of the instruction account payloads or instruction
+data.
+
+From the dApp devs perspective almost all changes are hidden inside the SDK.
```

Comment:
I am not sure we can hide the need for `sol_set_account_lamports` in the SDK. It might also be the case that CUs to update lamports will increase.